sdk/storage/Azure.Storage.Files.Shares/src/ShareFileClient.cs

```diff
@@ -5702,5 +6126,70 @@ public virtual Uri GenerateSasUri(ShareSasBuilder builder)
+        internal PartitionedDownloader<ShareFileRequestConditions, ShareFileDownloadInfo> GetPartitionedDownloader(
+            StorageTransferOptions transferOptions,
+            string operationName = null)
+            => new PartitionedDownloader<ShareFileRequestConditions, ShareFileDownloadInfo>(
+                GetPartitionedDownloaderBehaviors(this),
+                transferOptions,
+                operationName);
+
+        internal static PartitionedDownloader<ShareFileRequestConditions, ShareFileDownloadInfo>.Behaviors GetPartitionedDownloaderBehaviors(ShareFileClient client)
+        {
+            return new PartitionedDownloader<ShareFileRequestConditions, ShareFileDownloadInfo>.Behaviors
+            {
+                SingleDownload = async (range, args, rangeGetContentHash, async, cancellationToken)
+                    => await client.DownloadInternal(
+                        range,
+                        rangeGetContentHash,
+                        args,
+                        async,
+                        cancellationToken).ConfigureAwait(false),
+                CopyToAsync = async (result, destination, cancellationToken)
+                    =>
+                {
+                    using Stream source = result.Content;
+
+                    await source.CopyToAsync(
+                        destination,
+                        Constants.DefaultDownloadCopyBufferSize,
+                        cancellationToken)
+                        .ConfigureAwait(false);
+                },
+                CopyTo = (result, destination, cancellationToken)
+                    =>
+                {
+                    cancellationToken.ThrowIfCancellationRequested();
+                    result.Content.CopyTo(destination, Constants.DefaultDownloadCopyBufferSize);
+                    result.Content.Dispose();
+                },
+                CheckForEtagChange = (response, etag)
+                    =>
+                {
+                    // Make sure the ETag for this chunk matches the originally
+                    // recorded one
+                    if (etag != response.GetRawResponse().Headers.ETag)
+                    {
+                        ContentRange? contentRange = string.IsNullOrWhiteSpace(response.Value.Details.ContentRange)
+                            ? default
+                            : ContentRange.Parse(response.Value.Details.ContentRange);
+
+                        throw new ShareFileModifiedException(
+                            "File has been modified concurrently",
+                            client.Uri,
+                            etag,
+                            response.GetRawResponse().Headers.ETag.GetValueOrDefault(),
+                            new HttpRange(
+                                (long)contentRange?.Start.GetValueOrDefault(),
+                                contentRange?.End - contentRange?.Start));
+                    }
+                },
```

Comment:
To simplify the behaviors more. 
Consider accepting optional etag parameter in SingleDownload delegate instead.
In that case the downloader can run the first request with etag customer passed initially (if they pass) and then keep passing etag.
Then blobs can just pass it to the next chunk call.
In shares you can embed such check in the download delegate directly like we did here (ignore allowmodification flag it's openread specific) https://github.com/Azure/azure-sdk-for-net/blob/c87f56cd6184c318cd373ab7302618261d193170/sdk/storage/Azure.Storage.Files.Shares/src/ShareFileClient.cs#L2305-L2314 .